release: v0.11.0 (#1388)
apollographql/router: CHANGELOG.md

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# [0.11.0] - 2022-07-12

## ❗ BREAKING ❗

### Relax plugin api mutability ([PR #1340](https://github.com/apollographql/router/pull/1340) ([PR #1289](https://github.com/apollographql/router/pull/1289)

the `Plugin::*_service()` methods were taking a `&mut self` as argument, but since
they work like a tower Layer, they can use `&self` instead. This change
then allows us to move from Buffer to service factories for the query
planner, execution and subgraph services

> By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/1340 https://github.com/apollographql/router/pull/1289

Reviewer: we might wanna explain what this refactoring means for plugin writers, like they might have to rethink scopes (things that should be static might not be anymore)

Resources are almost like attributes but there are more globals. They are directly configured on the metrics exporter which means you'll always have these resources on each of your metrics. It could be pretty useful to set
a service name for example to let you more easily find metrics related to a specific service.

```yaml
telemetry:
  metrics:
    common:
      resources:
        # Set the service name to easily find metrics related to the apollo-router in your metrics dashboards
        service.name: "apollo-router"
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/1354

## 🐛 Fixes

### Fix fragment on interface without typename [PR #1371](https://github.com/apollographql/router/pull/1371)

When the subgraph doesn't return the typename and the type condition of a fragment is an interface, we should return the values if the entity implements the interface

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/1371

### Fix detection of an introspection query [PR #1370](https://github.com/apollographql/router/pull/1370)

> A query with at the root only one selection field equals to `__typename` must be considered as an introspection query

Reviewer: A query that only contains `__typename` at the root must be considered as an introspection query

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# [0.11.0] - 2022-07-12

## ❗ BREAKING ❗

### Relax plugin api mutability ([PR #1340](https://github.com/apollographql/router/pull/1340) ([PR #1289](https://github.com/apollographql/router/pull/1289)

the `Plugin::*_service()` methods were taking a `&mut self` as argument, but since
they work like a tower Layer, they can use `&self` instead. This change
then allows us to move from Buffer to service factories for the query
planner, execution and subgraph services.

**Services are now created on the fly at session creation, so if any state must be shared
between executions, it should be stored in an `Arc<Mutex<_>>` in the plugin and cloned
into the new service in the `Plugin::*_service()` methods**.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/1340 https://github.com/apollographql/router/pull/1289

## 🚀 Features

### Add support to add custom resources on metrics. [PR #1354](https://github.com/apollographql/router/pull/1354)

> Resources are almost like attributes but there are more globals. They are directly configured on the metrics exporter which means you'll always have these resources on each of your metrics. It could be pretty useful to set
> a service name for example to let you more easily find metrics related to a specific service.

Reviewer 2: ```suggestion
Resources are almost like attributes but more global. They are directly configured on the metrics exporter which means you'll always have these resources on each of your metrics.  This functionality can be used to, for example,
apply a `service.name` to metrics to make them easier to find in larger infrastructure, as demonstrated here:
```

Thoughts?

        service.name: "apollo-router"
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/1354

## 🐛 Fixes

### Fix fragment on interface without typename [PR #1371](https://github.com/apollographql/router/pull/1371)

When the subgraph doesn't return the typename and the type condition of a fragment is an interface, we should return the values if the entity implements the interface

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/1371

### Fix detection of an introspection query [PR #1370](https://github.com/apollographql/router/pull/1370)

A query that only contains __typename at the root must be considered as an introspection query

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/1370

### Accept nullable list as input [PR #1363](https://github.com/apollographql/router/pull/1363)

Do not throw a validation error when you give `null` for an input variable of type `[Int!]`.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/1363

## 🛠 Maintenance

> ### Replace Buffers of tower services with service factories([PR #1289](https://github.com/apollographql/router/pull/1289) [PR #1355](https://github.com/apollographql/router/pull/1355))

Reviewer 2: ```suggestion
### Replace Buffers of tower services with service factories ([PR #1289](https://github.com/apollographql/router/pull/1289) [PR #1355](https://github.com/apollographql/router/pull/1355))
```

formatting

tower services should be used by creating a new service instance for each new session
instead of going through a Buffer.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/1289  https://github.com/apollographql/router/pull/1355

> ### execute the query plan's first response directly ([PR #1357](https://github.com/apollographql/router/issues/1357))

Reviewer 2: ```suggestion
### Execute the query plan's first response directly ([PR #1357](https://github.com/apollographql/router/issues/1357))
```

> The query plan was entirely executed in a spawned task to prepare for the `@defer` implementation, but we can actually

Reviewer 2: ```suggestion
The query plan was previously executed in a spawned task to prepare for the `@defer` implementation, but we can actually
```
